### PR TITLE
Update `AddressForm` and related components to TypeScript

### PR DESCRIPTION
**Pull request by opr**
Converted `AddressForm`, `prepare-address-fields`, `BillingCountryInput`, `ShippingCountryInput`, `CountryInput`, `BillingStateInput`, `ShippingStateInput`, and `StateInput` to TypeScript.

This includes a cherry-picked commit from #4361 - so feel free to ignore the changes in `assets/js/settings/shared/utils.ts` for the purposes of this PR

### How to test the changes in this Pull Request:

1. Smoke test cart and checkout functionality.
2. Ensure that disallowed countries are not selectable in the checkout form.
3. Ensure address can be submitted and it is sent to the API correctly and stored correctly in the order.

### CODE REVIEW

File: assets/js/settings/blocks/constants.ts
Comment (Aljullu) [inline assets/js/settings/blocks/constants.ts]: `AllowedCountries` shouldn't be capitalized:
```suggestion
export const ALLOWED_COUNTRIES = getSetting( 'allowedCountries', {} ) as Record<
```

**Review by Aljullu — APPROVED** [review]
Code changes look good and this is testing good. I needed to make a small fix to get countries to show up, though. I left a comment inline below. :point_down: 

Pre-approving because once that's fixed, it should be good to merge.

File: assets/js/settings/blocks/constants.ts
Comment (opr) [reply assets/js/settings/blocks/constants.ts]: Thanks Albert, good spot 🕵🏼

**Pull request merged**
